### ISSUE TITLE
PoC: Combine 'In Discussion' and 'Voting' tabs

### ISSUE THREAD

**Issue by lukebp** [decred/politeiagui#2394]
**Problem**

We removed indicators on the home page tabs (`In Discussion`, `Voting`, `Approved`, etc) that displayed the proposal counts.  This was done for a good reason, but now there is unnecessary friction required when a user wants to see if there are any proposals being voted on.  The page defaults to the `In Discussion` tab.  The user must now click to the voting tab to see if there are any proposals being voted on.

**Solution**

Lets put together a proof of concept that combines the `In Discussion` tab and the `Voting` tab into a single tab.  The home page will default to this tab.  You can display the proposals in the same order that they appear in the `ticketvote` `InventoryReply`.  `Started` would appear first then `Authorized` then `Unauthorized`.

The name `Active` comes to mind for this tab, but I think we should reserve `Active` to mean the proposal is actively being billed against (see the proposal lifecycle section of the [pi development proposal](https://proposals.decred.org/record/91cfcc8)).  Any ideas for this tab name?

Full matrix discussion
```
davecgh
Is there a technical reason the counts were removed from the tabs, btw?  I miss
them!  It was easy to quickly see the status of proposals from just the tabs.
Now I have to click through to voting to see if it's started/still voting.

lukebp
the old politeia had a route that returned all tokens in the inventory
categorized vote status. The gui just keyed off of those to display the tab
count. The new politeia paginates that response. The gui can't key off of it
anymore.

a new route could be added that returns the proposal counts, but generally
speaking, I'm not a fan of the tab count UI. Its only helpful when there are a
small number of proposals in the system. Something like Approved (323) doesn't
really provide you with any valuable information. Tab counts don't scale well
as a UI feature.

I think the UI will likely evolve and move away from the tabs as the number of
proposals grow and as more proposal metadata is required, ex. adding the domain
to the proposal (dev, marketing, etc). This is more of a shower thought at this
point, but I think a design worth exploring would be for the home page to
display proposals based on recent activity (in discussion, voting, recently
ended) and then you would have domain based subreddits that you would navigate
to if you wanted to look back through the proposal history.

maybe combining the In Discussion and Voting tabs into a single tab would be a
good idea so that you don't have to click through to see all the proposals that
are relevant for the user at that point in time

jy-p
tag the entries by color to denote which ones are discussion versus voting

lukebp
we can explore combining In Discussion and Voting into just Active and have the
list items differentiate between In Discussion and Voting using either color or
the status tag that's currently in the top right corner
```

**Comment by lukebp** [comment]
Possible naming choice for the tabs: `Under Review`, `Approved`, `Rejected`, `Abandoned`

`Under Review` are proposals that are being discussed and proposals that are being voted on.

**Comment by victorgcramos** [comment]
Will the backend return the In Discussion and Voting lists separate when fetching the token inventory? Because if not, we will need to fetch all tokens first, so we don't mess it up when paginating.

**Comment by lukebp** [comment]
The backend returns the inventory by record status and by vote status.  Its up to the client to figure out how it wants to display them.

**Comment by victorgcramos** [comment]
Okay, so I think we would have some edge-cases when implementing this:

1. If we have >20 proposals under discussion and >20 proposals being voted. The first token inventory call (page 1) will return the first 20 tokens for each. The list will load, for example `[...first 20 voting, ...first 20 under discussion]`. What will happen if we load more? The list will get messed up... Either we do it like `[...first 20 voting, ...second 20 voting, ...first 20 under discussion, ...second 20 under discussion]` or `[...fist 20 voting, ...first 20 under discussion, ...second 20 voting, ...second 20 discussion]`.

2. The longer the list is, the harder it will be to find some proposal. In terms of UX, scrolling down a list every time you want to check for some proposal update can be very painful when lists are too long. Imagine a user wants to check for new proposals. He/she would have to scroll all the voting proposals before reaching the target.

3. If the token inventory request returns list ordered by date, when mixing 2 lists, we would also lose the date reference and break the existing pattern on GUI.

In conclusion, I think that this feature should be implemented along with some filtering for users that want to display it separate, in order to avoid the mentioned list confusion, specially if the lists have 2 or more pages.

**Comment by lukebp** [comment]
1. You fetch the specific vote status until there you get a page < pageSize, which means there are no more.  This is not an issue.

2. This would very rarely, if ever, happen.  This is not an issue. 

3. You don't mix the two lists.  You display one on top of the other.  This doesn't break anything.

None of these are blocking issues.

**Comment by victorgcramos** [comment]
> You don't mix the two lists. You display one on top of the other. This doesn't break anything.

Both "top" and "bottom" lists will have their sizes changed if there are more than 20 proposals. The `/records` endpoint will only return up to 10 proposals. We first need to fetch the inventory and then fetch the records, right?

When we fetch the inventory, the max page size is 20, so if we have more than 20 proposals under discussion, the first `/inventory` call won't be complete. We would have to do another request to fetch the remaining tokens. But the first call, we don't specify the status, since it returns the first 20 for each status anyway. In this case, we would have maximum `20 under discussion` and `20 active voting`. So the first list display, on the `Under Review` tab you mentioned, will list the first `20 active voting` followed by the `20 under discussion`:
```
- active voting #1
- active voting #2
- ...
- active voting #20
- under discussion #1
- under discussion #2
- ...
- under discussion #20
```

What will happen when we fetch the 21st for the active voting? We have 2 possibilities:
1. ```
    - active voting #1
    - active voting #2
    - ...
    - active voting #20
    - active voting #21 <-- this will appear here
    - under discussion #1
    - under discussion #2
    - ...
    - under discussion #20
   ```
   
2. ```
    - active voting #1
    - active voting #2
    - ...
    - active voting #20
    - under discussion #1
    - under discussion #2
    - ...
    - under discussion #20
    - active voting #21 <-- this will appear here
   ```
  
For both scenarios, the list will be mixed. In terms of UX, it's expected to display the new items below the list, usually where the user scroll height is. If we follow the pattern, we will have the lists mixed by status. If we don't, users will face difficulties to find the loaded proposals, meaning he/she would have to scroll the whole list again to find the recently fetched proposals.

Or we can load the whole inventory first, order it, and then fetch it accordingly, leading us to longer loading time at an expensive cost.

**Comment by lukebp** [comment]
I'll expand a bit on how this would work. The `Under Review` tab will list the proposals with the vote status `started`, `authorized`, `unauthorized`.  They will be listed in that order.  `started` will be listed first, then `authorized`, then `unauthorized`.

1. The user navigates to the proposals site.  The site defaults to the `Under Review` tab.
2. The gui fetches the vote inventory.
3. The gui displays the `started`, `authorized`, and `unauthorized` proposals in that order.  If the page of tokens in one of the vote statuses contains less than 20 tokens (the inventory page size) then you know that there are no more proposals with that vote status.  If a vote status contains 20 tokens then you know that you need to keep fetching additional pages of tokens.  

For example, let's say that there are 5 proposals under vote, 23 authorized, and 2 unauthorized.  The gui would display the 5 proposals under vote first.  If the user starts scrolling down, the gui would display the authorized proposals until it reached the end of the first page of returned tokens.  This will be 20 tokens.  Because a full page of tokens was returned for authorized proposals, the gui knows that it needs to keep fetching additional pages of authorized tokens.  The gui fetches page 2 of the authorized vote inventory.  The reply contains the additional 3 tokens.  The gui now knows that there are no more authorized proposals since the returned page contained less than a full page of tokens (20 tokens).  The gui moves on to fetching the unauthorized proposals.

@victorgcramos does this answer your questions and alleviate your concerns?

**Comment by victorgcramos** [comment]
But what If I just want to see the `unauthorized` ones? I'd have to wait for the token fetching pipeline + record fetch every time? Basically, users will only notice unauthorized proposals if and only if he/she scrolls the entire list, right?

I'm worried about this because once a proposal gets authorized or under vote, it cannot be edited anymore. The unauthorized status is where the proposal author can get some feedback and edit the proposal to clarify the stakeholders questions, and stakeholders shouldn't face this loading time barrier in order to participate on some proposal's discussion.

**Comment by lukebp** [comment]
> But what If I just want to see the unauthorized ones? I'd have to wait for the token fetching pipeline + record fetch every time? Basically, users will only notice unauthorized proposals if and only if he/she scrolls the entire list, right?

Then you keep scrolling down, same as if there are more than 5 under discussion proposals now.  This is not an issue imo.  The whole reason we're making this change is because there's very rarely more than a small handful of proposals under review at any single time.  If this changes in the future then we can adapt the implementation, but I don't think it will.

Another benefit of this is that it would allow us to remove the record inventory calls.  The tabs would map directly to the vote statuses:
Under Review: `aunauthorized`, `authorized`, `voting`
Started: `started`
Approved: `approved`
Rejected: `rejected`
Abandoned: `ineligible`

**Comment by victorgcramos** [comment]
The problem here is not on the technical development domain, it's mostly an UX concern. On most screens, users only face 4 loaded proposals. If the first 4 don't change, the user won't notice any new proposals. What if the 8 first records don't change either? What about 12? 16?

**Comment by lukebp** [comment]
> Since the token inventory request is ordered by submission date

They're ordered by the timestamp of their most recent status change.

I understand what you're saying and I don't agree.  Your arguments are based on situations where there are a large number of concurrent proposals.  The entire reason we're having this discussion is because large numbers of concurrent proposals do not  happen in practice and it has lead to a bad UX on the existing site.  Proposals are rare.  If this changes to a significant degree in the future then we can reasses.  We should optimize UX for what occurs in practice, not for a theoretical case of 15 concurrent proposals.

**Comment by lukebp** [comment]
The degree that this simplifies the exsiting implementation by removing the record inventory calls should also not be underestimated.  This is a big benefit.

**Comment by victorgcramos** [comment]
I assume the `unauthorized` list is the most important from the stakeholder PoV, and it will probably be the longest among the regarded `Under Review` list. The larger and the "most updated" ones. Before any status changes, proposals must pass through the `in discussion` stage depending on stakeholders discussion. This stage is where we define whether a proposal should be authorized or abandoned, and it's the only stage where non-author/admin inputs are allowed.

The `unauthorized` list, IMO, should be the first one to be fetched. But if we fetch it first, since it tend to be a longer list, users won't notice which proposals are `under vote` or `authorized` unless they scroll the whole list.

I don't think we need to have a huge number of proposals `authorized` + `voting` to "hide" the `unauthorized` ones. If we have at least 8 proposals in this union, and I think this is enough for users to ignore the rest of the list and assume there are no new proposals under discussion.

Politeia is mostly a discussion platform, and we keep the discussion harder to access is not a good improvement in terms of usability. We win on code performance + implementation, which is a bonus only us devs can notice, but we lose the UX, which is an onus for the average Joe.

But yes, like I said on my first comment, it's an edge case.

**Comment by lukebp** [comment]
This discussion has run its course.  We're going to keep things simple and have the tabs correspond to vote statuses.

Under Review: `anauthorized`, `authorized`, `voting`
Started: `started`
Approved: `approved`
Rejected: `rejected`
Ineligible: `ineligible`

The exact gui behavior has been outlined in this comment.

https://github.com/decred/politeiagui/issues/2394#issuecomment-853964068

**Comment by victorgcramos** [comment]
We don't have the "voting" status defined here, by the way.
https://github.com/decred/politeia/blob/466bdbc0615d7bae7a8067823cf309da53048c56/politeiawww/api/ticketvote/v1/v1.go#L209-L214

And the "In Discussion" tab is currently comprising both authorized and unauthorized ones:
https://github.com/decred/politeiagui/blob/c63222fcf58bffd483ca6079f6bc1f3efd9238f6/src/actions/api.js#L569-L572).

**Comment by vibros68** [comment]
After to some testings with proposals. I see the current logic is fetch proposals without status condition and do the filter at frontend side. So we can simply combine `In Discussion` and `Voting` to be `Under Review` and do some fix on the filter. 

But with my experience I prefer to do the filtering at backend side and the backend should support to fetch multi statuses in a request

**Comment by victorgcramos** [comment]
The current proposal fetching logic uses the proposal vote status after an initial fetch. Records length will always grow up, so filtering on the frontend side isn't a good solution in terms of scalability. You should use the frontend just to combine the lists, not for filtering.

Token inventory calls are also paginated, so you should fetch the next token inventory page only when required, i.e. when you have no more proposals to load from the inventory, and the inventory length for some status is a multiple of 20.

**Comment by lukebp** [comment]
> But with my experience I prefer to do the filtering at backend side and the backend should support to fetch multi statuses in a request

The backend is designed to be generic in order to support multiple applications being built on top.  Its the responsibility of the client to filter/combine the generic responses in a manner that is suitable for the client.

**Comment by vibros68** [comment]
I am working on it.

**Comment by victorgcramos** [comment]
Hey @vibros68 noticed some issues regarding token inventory pagination. #2406 removed the pagination by status, causing to not work properly. I'm fixing this on this PR, and I think you should wait until it gets fixed so you can understand how the pagination works for the token inventory.
